File: src/CandidatosCore.Core/Notificaciones/SuscriptorNotificacion.cs
namespace CandidatosCore.Core.Notificaciones;

using CandidatosCore.Core.Historial;

public sealed class SuscriptorNotificacion : ObservadorCambio
{
    public const string Todos = "*";

    public string Destinatario { get; }
    public IReadOnlySet<string> EventosDeInteres { get; }

    public bool conComentario { get; }

    public SuscriptorNotificacion(string destinatario, IEnumerable<string> eventosDeInteres, bool? conComentario)
    {
        Destinatario = destinatario;
        EventosDeInteres = new HashSet<string>(eventosDeInteres);
        this.conComentario = conComentario ?? false;
    }

    public void Actualizar(RegistroCambio cambio)
    {
        if (!EventosDeInteres.Contains(Todos) && !EventosDeInteres.Contains(cambio.EstadoNuevo.Nombre))
            return;

        string descripcion = conComentario ? $"{cambio.Comentario}" : null;

        EmailService.Enviar(Destinatario, DescribirCambio(cambio), comentario);
    }

    private static string DescribirCambio(RegistroCambio cambio) => cambio.Tipo switch
    {
        TipoCambio.Deshacer => $"{cambio.Candidato.Nombre}: se deshizo el último cambio, ahora en {cambio.EstadoNuevo.Nombre}",
        TipoCambio.Rechazo => $"{cambio.Candidato.Nombre} fue rechazado",
        TipoCambio.Avance => $"{cambio.Candidato.Nombre} pasó a {cambio.EstadoNuevo.Nombre}",
        _ => $"Error al describir el cambio: {cambio.Tipo}"
    };
}

File: src-ts/CandidatosCore.Core/Notificaciones/SuscriptorNotificacion.ts
import type { RegistroCambio } from "../Historial/RegistroCambio";
import { TipoCambio } from "../Historial/TipoCambio";
import { EmailService } from "./EmailService";
import type { ObservadorCambio } from "./ObservadorCambio";

export class SuscriptorNotificacion implements ObservadorCambio {
    public static readonly Todos = "*";

    public readonly destinatario: string;
    public readonly eventosDeInteres: Set<string>;
    public readonly conComentario: boolean;

    constructor(
        destinatario: string,
        eventosDeInteres: Iterable<string>,
        conComentario?: boolean
    ) {
        this.destinatario = destinatario;
        this.eventosDeInteres = new Set(eventosDeInteres);
        this.conComentario = conComentario ?? false;
    }

    public actualizar(cambio: RegistroCambio): void {
        if (
            !this.eventosDeInteres.has(SuscriptorNotificacion.Todos) &&
            !this.eventosDeInteres.has(cambio.estadoNuevo.nombre)
        ) {
            return;
        }

        const comentario = this.conComentario ? `${cambio.comentario ?? ""}` : undefined;

        EmailService.enviar(this.destinatario, SuscriptorNotificacion.describirCambio(cambio), comentario);
    }

    private static describirCambio(cambio: RegistroCambio): string {
        switch (cambio.tipo) {
            case TipoCambio.Deshacer:
                return `${cambio.candidato.nombre}: se deshizo el último cambio, ahora en ${cambio.estadoNuevo.nombre}`;
            case TipoCambio.Rechazo:
                return `${cambio.candidato.nombre} fue rechazado`;
            case TipoCambio.Avance:
                return `${cambio.candidato.nombre} pasó a ${cambio.estadoNuevo.nombre}`;
            default:
                return `Error al describir el cambio: ${cambio.tipo}`;
        }
    }
}
